DataStructures/ArrayBasedQueue.cs

```diff
@@ -0,0 +1,101 @@
+using System;
+using System.Collections.Generic;
+
+namespace DataStructures
+{
+    /// <summary>
+    /// Implementation of an array based queue. FIFO style.
+    /// </summary>
+    /// <typeparam name="T">Generic Type.</typeparam>
+    public class ArrayBasedQueue<T>
+    {
+        private readonly T[] queue;
+        private int currentIndex;
+        private readonly int capacity;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ArrayBasedQueue{T}"/> class.
+        /// </summary>
+        public ArrayBasedQueue(int capacity)
+        {
+            this.capacity = capacity;
+            queue = new T[capacity];
+        }
+
+        /// <summary>
+        /// Clears the queue.
+        /// </summary>
+        public void Clear()
+        {
+            currentIndex = 0;
+        }
+
+        /// <summary>
+        /// Returns the first item in the queue and removes it from the queue.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">Thrown if the queue is empty.</exception>
+        public T Dequeue()
+        {
+            if (IsEmpty())
+            {
+                throw new InvalidOperationException("There are no items in the queue.");
+            }
+
+            var item = queue[0];
+
+            for (int i = 0; i < currentIndex - 1; i++)
+            {
+                queue[i] = queue[i + 1];
```

Comment:
Unfortunately, this is not the way to go. In queue Enqueue and Dequeue operations must be performed in O(1) time, but this is O(n). See [this](https://www.geeksforgeeks.org/circular-queue-set-1-introduction-array-implementation/) and [this](https://en.wikipedia.org/wiki/Circular_buffer) for some hints

Reply:
Yeah, sorry for being lazy 😄